lib/marisa/grimoire/vector/bit-vector.cc

```diff
@@ -784,22 +790,45 @@ void BitVector::build_index(const BitVector &bv,
       }
     }
 
-    if (bv[i]) {
-      if (enables_select1 && ((num_1s % 512) == 0)) {
-        select1s_.push_back(static_cast<UInt32>(i));
+    const Unit unit = bv.units_[unit_id];
+    // push_back resizes with 0, so the high bits of the last unit are 0 and
+    // do not affect the 1s count.
+    const std::size_t unit_num_1s = PopCount::count(unit);
+
+    if (enables_select0) {
+      // num_0s is somewhat move involved to compute, so only do it if we
+      // need it.  The last word has zeros in the high bits, so that needs
+      // to be accounted for when computing the unit_num_0s from unit_num_1s.
+      const std::size_t bits_remaining = num_bits - bit_id;
+      const std::size_t unit_num_0s =
+          std::min<std::size_t>(bits_remaining, MARISA_WORD_SIZE) - unit_num_1s;
+
+      const std::size_t zero_bit_id = -num_0s % 512;
+      if (unit_num_0s > zero_bit_id) {
+        // select0s_ is UInt32, but select_bit returns size_t, so cast to
+        // suppress narrowing conversion warning.  push_back checks the
+        // size, so there is no truncation here.
+        select0s_.push_back(
+            static_cast<UInt32>(select_bit(zero_bit_id, bit_id, ~unit)));
```

Comment:
32-bit doesn't define `select_bit` unless `MARISA_USE_SSE2`, in which case it takes 2 `UInt32` args. 

Don't merge this because I need to fix it.

Reply:
Fixed.